File: slopsmith/lib/audio.py
"""Audio extraction and conversion for Rocksmith CDLC."""

import os
import shutil
import subprocess
from pathlib import Path


def _vgmstream_cmd() -> str | None:
    """Return the path to vgmstream-cli if available."""
    return shutil.which("vgmstream-cli")


def _ffmpeg_cmd() -> str | None:
    """Return the path to ffmpeg if available."""
    return shutil.which("ffmpeg")


def encode_wav_to_ogg(
    wav_path: str | Path,
    ogg_path: str | Path,
    quality: int = 5,
    ffmpeg: str | None = None,
) -> None:
    """Encode a WAV (or any ffmpeg-decodable audio) to Ogg/Vorbis.

    Prefers the ``libvorbis`` encoder; falls back to ffmpeg's built-in
    (experimental) ``vorbis`` encoder when the ffmpeg build doesn't link
    libvorbis (e.g. stock Homebrew ffmpeg). Raises RuntimeError on failure.
    """
    ff = ffmpeg or _ffmpeg_cmd() or "ffmpeg"
    ogg_path = Path(ogg_path)
    ogg_path.parent.mkdir(parents=True, exist_ok=True)

    # Primary: libvorbis
    r = subprocess.run(
        [ff, "-y", "-loglevel", "error", "-i", str(wav_path),
         "-c:a", "libvorbis", "-q:a", str(quality), str(ogg_path)],
        capture_output=True,
    )
    if r.returncode == 0 and ogg_path.exists() and ogg_path.stat().st_size > 0:
        return

    # Fallback: native vorbis (experimental -> needs -strict -2)
    if ogg_path.exists() and ogg_path.stat().st_size == 0:
        ogg_path.unlink()
    r2 = subprocess.run(
        [ff, "-y", "-loglevel", "error", "-i", str(wav_path),
         "-c:a", "vorbis", "-strict", "-2", "-q:a", str(quality), str(ogg_path)],
        capture_output=True,
    )
    if r2.returncode != 0 or not ogg_path.exists() or ogg_path.stat().st_size < 100:
        raise RuntimeError(
            f"ffmpeg OGG/Vorbis encode failed for {Path(wav_path).name}: "
            f"{r2.stderr.decode(errors='replace')[-400:]}"
        )


def find_wem_files(extracted_dir: str) -> list[str]:
    """Find WEM audio files, sorted largest first (full song before preview)."""
    wem_files = list(Path(extracted_dir).rglob("*.wem"))
    wem_files.sort(key=lambda p: p.stat().st_size, reverse=True)
    return [str(f) for f in wem_files]


def convert_wem(wem_path: str, output_base: str) -> str:
    """
    Convert a WEM file to a playable format.
    Returns path to the converted audio file.
    """
    # Try vgmstream-cli → WAV → MP3 (best browser compatibility)
    if shutil.which("vgmstream-cli"):
        wav = output_base + ".wav"
        r = subprocess.run(
            ["vgmstream-cli", "-o", wav, wem_path], capture_output=True
        )
        if r.returncode == 0 and os.path.exists(wav) and os.path.getsize(wav) > 0:
            if shutil.which("ffmpeg"):
                mp3 = output_base + ".mp3"
                r2 = subprocess.run(
                    ["ffmpeg", "-y", "-i", wav, "-b:a", "192k", mp3],
                    capture_output=True,
                )
                if r2.returncode == 0 and os.path.exists(mp3):
                    os.remove(wav)
                    return mp3
            return wav

    # Try ffmpeg directly (some builds handle Wwise)
    if shutil.which("ffmpeg"):
        mp3 = output_base + ".mp3"
        r = subprocess.run(
            ["ffmpeg", "-y", "-i", wem_path, "-b:a", "192k", mp3],
            capture_output=True,
        )
        if r.returncode == 0 and os.path.exists(mp3) and os.path.getsize(mp3) > 0:
            return mp3

        # Try WAV output as fallback
        wav = output_base + ".wav"
        r = subprocess.run(
            ["ffmpeg", "-y", "-i", wem_path, wav],
            capture_output=True,
        )
        if r.returncode == 0 and os.path.exists(wav) and os.path.getsize(wav) > 0:
            return wav

    # Try ww2ogg
    if shutil.which("ww2ogg"):
        ogg = output_base + ".ogg"
        r = subprocess.run(
            ["ww2ogg", wem_path, "-o", ogg], capture_output=True
        )
        if r.returncode == 0 and os.path.exists(ogg) and os.path.getsize(ogg) > 0:
            return ogg

    raise RuntimeError(
        "No WEM audio decoder found. Install vgmstream-cli:\n"
        "  Manjaro/Arch:  yay -S vgmstream-cli-bin\n"
        "  Or build from: github.com/vgmstream/vgmstream"
    )

File: rust4slopsmith/src/engine/audio.rs
//! Audio extraction/conversion for Rocksmith CDLC. Port of `lib/audio.py` —
//! pure subprocess orchestration (ffmpeg / vgmstream-cli / ww2ogg), no binary
//! parsing of its own.

use std::path::{Path, PathBuf};
use std::process::Command;

/// `shutil.which` — find a binary on PATH.
fn which(cmd: &str) -> Option<PathBuf> {
    let paths = std::env::var_os("PATH")?;
    for dir in std::env::split_paths(&paths) {
        let candidate = dir.join(cmd);
        if candidate.is_file() {
            return Some(candidate);
        }
    }
    None
}

fn ffmpeg_cmd() -> Option<PathBuf> {
    which("ffmpeg")
}

/// Encode a WAV (or any ffmpeg-decodable audio) to Ogg/Vorbis. Prefers
/// `libvorbis`; falls back to the experimental native `vorbis` encoder when
/// the ffmpeg build doesn't link libvorbis (e.g. stock Homebrew). Mirrors
/// `encode_wav_to_ogg` (audio.py:19-56).
pub fn encode_wav_to_ogg(wav_path: &Path, ogg_path: &Path, quality: i32, ffmpeg: Option<&Path>) -> anyhow::Result<()> {
    let ff = ffmpeg
        .map(|p| p.to_path_buf())
        .or_else(|| ffmpeg_cmd())
        .unwrap_or_else(|| PathBuf::from("ffmpeg"));
    if let Some(parent) = ogg_path.parent() {
        std::fs::create_dir_all(parent)?;
    }
    let q = quality.to_string();

    // Primary: libvorbis.
    let r = Command::new(&ff)
        .args(["-y", "-loglevel", "error"])
        .arg("-i").arg(wav_path)
        .args(["-c:a", "libvorbis"])
        .arg("-q:a").arg(&q)
        .arg(ogg_path)
        .output();
    if let Ok(out) = &r {
        if out.status.success()
            && ogg_path.exists()
            && ogg_path.metadata().map(|m| m.len() > 0).unwrap_or(false)
        {
            return Ok(());
        }
    }

    // Fallback: native vorbis (experimental → -strict -2).
    if ogg_path.exists() && ogg_path.metadata().map(|m| m.len() == 0).unwrap_or(false) {
        std::fs::remove_file(ogg_path).ok();
    }
    let r2 = Command::new(&ff)
        .args(["-y", "-loglevel", "error"])
        .arg("-i").arg(wav_path)
        .args(["-c:a", "vorbis", "-strict", "-2"])
        .arg("-q:a").arg(&q)
        .arg(ogg_path)
        .output()?;
    let ok = r2.status.success()
        && ogg_path.exists()
        && ogg_path.metadata().map(|m| m.len() >= 100).unwrap_or(false);
    if !ok {
        let stderr = String::from_utf8_lossy(&r2.stderr);
        anyhow::bail!(
            "ffmpeg OGG/Vorbis encode failed for {}: {}",
            wav_path.file_name().and_then(|s| s.to_str()).unwrap_or("?"),
            stderr.chars().rev().take(400).collect::<String>().chars().rev().collect::<String>()
        );
    }
    Ok(())
}

/// Find WEM audio files, sorted largest first (full song before preview).
/// Mirrors `find_wem_files` (audio.py:59-63).
pub fn find_wem_files(extracted_dir: &Path) -> Vec<PathBuf> {
    let mut wem: Vec<(u64, PathBuf)> = Vec::new();
    for e in walkdir::WalkDir::new(extracted_dir).into_iter().filter_map(|e| e.ok()) {
        if e.path().extension().and_then(|x| x.to_str()) == Some("wem") {
            if let Ok(meta) = e.metadata() {
                wem.push((meta.len(), e.path().to_path_buf()));
            }
        }
    }
    wem.sort_by(|a, b| b.0.cmp(&a.0));
    wem.into_iter().map(|(_, p)| p).collect()
}

/// Convert a WEM file to a playable format. Returns the converted file path.
/// Tries vgmstream-cli → WAV → MP3, then ffmpeg, then ww2ogg. Mirrors
/// `convert_wem` (audio.py:66-121).
pub fn convert_wem(wem_path: &Path, output_base: &Path) -> anyhow::Result<PathBuf> {
    let wav = output_base.with_extension("wav");
    let mp3 = output_base.with_extension("mp3");
    let ogg = output_base.with_extension("ogg");

    if let Some(vgm) = which("vgmstream-cli") {
        let r = Command::new(&vgm).arg("-o").arg(&wav).arg(wem_path).output();
        if let Ok(out) = &r {
            if out.status.success() && wav.exists() && wav.metadata().map(|m| m.len() > 0).unwrap_or(false) {
                if let Some(ff) = ffmpeg_cmd() {
                    let r2 = Command::new(&ff).args(["-y", "-i"]).arg(&wav).args(["-b:a", "192k"]).arg(&mp3).output();
                    if let Ok(out2) = &r2 {
                        if out2.status.success() && mp3.exists() {
                            std::fs::remove_file(&wav).ok();
                            return Ok(mp3);
                        }
                    }
                }
                return Ok(wav);
            }
        }
    }

    if let Some(ff) = ffmpeg_cmd() {
        let r = Command::new(&ff).args(["-y", "-i"]).arg(wem_path).args(["-b:a", "192k"]).arg(&mp3).output();
        if let Ok(out) = &r {
            if out.status.success() && mp3.exists() && mp3.metadata().map(|m| m.len() > 0).unwrap_or(false) {
                return Ok(mp3);
            }
        }
        let r = Command::new(&ff).args(["-y", "-i"]).arg(wem_path).arg(&wav).output();
        if let Ok(out) = &r {
            if out.status.success() && wav.exists() && wav.metadata().map(|m| m.len() > 0).unwrap_or(false) {
                return Ok(wav);
            }
        }
    }

    if let Some(ww) = which("ww2ogg") {
        let r = Command::new(&ww).arg(wem_path).arg("-o").arg(&ogg).output();
        if let Ok(out) = &r {
            if out.status.success() && ogg.exists() && ogg.metadata().map(|m| m.len() > 0).unwrap_or(false) {
                return Ok(ogg);
            }
        }
    }

    anyhow::bail!(
        "No WEM audio decoder found. Install vgmstream-cli:\n  \
         Manjaro/Arch: yay -S vgmstream-cli-bin\n  \
         Or build from: github.com/vgmstream/vgmstream"
    )
}
